recipe-client-addon/lib/RecipeRunner.jsm

```diff
@@ -39,88 +39,152 @@ Cu.importGlobalProperties(["fetch"]);
+    if (this.enabled && (devMode || firstRun)) {
       await this.run();
-      prefs.setBoolPref(FIRST_RUN_PREF, false);
     }
+    if (firstRun) {
+      Services.prefs.setBoolPref(FIRST_RUN_PREF, false);
+    }
+  },
 
+  enable() {
+    if (this.enabled) {
+      return;
+    }
+    this.enabled = true;
     this.registerTimer();
   },
 
-  registerTimer() {
-    this.updateRunInterval();
-    CleanupManager.addCleanupHandler(() => timerManager.unregisterTimer(TIMER_NAME));
-
-    // Watch for the run interval to change, and re-register the timer with the new value
-    prefs.addObserver(RUN_INTERVAL_PREF, this);
-    CleanupManager.addCleanupHandler(() => prefs.removeObserver(RUN_INTERVAL_PREF, this));
+  disable() {
+    if (!this.enabled) {
+      return;
+    }
+    this.enabled = false;
+    this.unregisterTimer();
   },
 
-  checkPrefs() {
-    // Only run if Unified Telemetry is enabled.
-    if (!Services.prefs.getBoolPref("toolkit.telemetry.unified")) {
-      log.info("Disabling RecipeRunner because Unified Telemetry is disabled.");
-      return false;
+  /** Watch for prefs to change, and call this.observer when they do */
+  watchPrefs() {
+    for (const pref of PREFS_TO_WATCH) {
+      Services.prefs.addObserver(pref, this);
     }
 
-    if (!prefs.getBoolPref("enabled")) {
-      log.info("Recipe Client is disabled.");
-      return false;
-    }
+    CleanupManager.addCleanupHandler(this.unwatchPrefs.bind(this));
+  },
 
-    const apiUrl = prefs.getCharPref("api_url");
-    if (!apiUrl || !apiUrl.startsWith("https://")) {
-      log.error(`Non HTTPS URL provided for extensions.shield-recipe-client.api_url: ${apiUrl}`);
-      return false;
+  unwatchPrefs() {
+    for (const pref of PREFS_TO_WATCH) {
+      Services.prefs.removeObserver(pref, this);
     }
-
-    return true;
   },
 
+  /** When prefs change, this is fired */
   observe(subject, topic, data) {
     switch (topic) {
-      case PREF_CHANGED_TOPIC:
-        this.observePrefChange(data);
+      case PREF_CHANGED_TOPIC: {
+        const prefName = data;
+
+        switch (prefName) {
+          case RUN_INTERVAL_PREF:
+            this.updateRunInterval();
+            break;
+
+          // explicit fall-through
+          case TELEMETRY_ENABLED_PREF:
+          case SHIELD_ENABLED_PREF:
+          case API_URL_PREF:
+            this.checkPrefs();
+            break;
+
+          default:
+            log.debug(`Observer fired with unexpected pref change: ${prefName}`);
+        }
+
         break;
+      }
     }
   },
 
-  /**
-   * Watch for preference changes from Services.pref.addObserver.
-   */
-  observePrefChange(prefName) {
-    if (prefName === RUN_INTERVAL_PREF) {
-      this.updateRunInterval();
+  checkPrefs() {
+    let shouldEnable = true;
+
+    // Only run if Unified Telemetry is enabled.
+    if (!Services.prefs.getBoolPref(TELEMETRY_ENABLED_PREF)) {
+      log.debug("Disabling RecipeRunner because Unified Telemetry is disabled.");
```

Comment:
I like that this states what it's doing, can we update the other logs to also state that they're disabling Shield?

Reply:
I wonder if having multiple log messages claiming to be disabling the recipe runner would be confusing to, say, QA. Dunno if it's better to disable immediately and return, or list out the reasons like it is currently.